### PR TITLE
Create scraper for CourseBook data

### PR DESCRIPTION
**Pull request by CharlieMahana**
Closes #22.

### CODE REVIEW

File: scraper/scraper.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        # create reportmonkey downloadURL from downloadID
+        downloadURL = 'https://coursebook.utdallas.edu/reportmonkey/cb11-export/{0}/{0}/json'.format(downloadID)
+
+        # specify reportmonkey request header
+        reportmonkeyHeader = {
+            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
+            'Accept-Encoding': 'gzip, deflate, br',
+            'Accept-Language': 'en-US,en;q=0.9',
+            'Connection': 'keep-alive',
+            'Cookie': 'PTGSESSID={}'.format(self.ptgsessid),
```
Comment (WillieCubed) [inline scraper/scraper.py:141]: f-strings are cooler.
```suggestion
            'Cookie': f'PTGSESSID={self.ptgsessid}',
```

File: scraper/scraper.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+            'Sec-Fetch-Mode': 'navigate',
+            'Sec-Fetch-Site': 'same-site',
+            'Sec-Fetch-User': '?1',
+            'Sec-GPC': '1',
+            'Upgrade-Insecure-Requests': '1',
+            'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.81 Safari/537.36',
+        }
+
+        # send coursemonkey download request for json
+        # self.logger.info('Sending ReportMonkey query: {0}'.format(query))
```
Comment (WillieCubed) [inline scraper/scraper.py:153]: Remove commented out code when merging to `develop`. If it's important, add it to another branch.

File: scraper/scraper.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+
+    def getSearchSpecifications(self) -> dict:
+        self.logger.info('Obtaining CourseBook search specifications')
+        specifications = {}
+
+        # obtain coursebook index page
+        self.logger.info('Getting CourseBook index page')
+        try:
+            r = requests.get('https://coursebook.utdallas.edu')
```
Comment (WillieCubed) [inline scraper/scraper.py:45]: Could you extract the CourseBook URL to a top-level variable and replace it anywhere it is used?

File: scraper/scraper.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+            if downloadLink[0:14] == '/reportmonkey/':
+                downloadID =  downloadLink.split('/')[-1]
+            else:
+                raise Exception()
+        except:
+            self.logger.error('Failed to parse download link to {0} item(s) from query: {1}'.format(numItems, query))
+            return {}
+
+        # create reportmonkey downloadURL from downloadID
+        downloadURL = 'https://coursebook.utdallas.edu/reportmonkey/cb11-export/{0}/{0}/json'.format(downloadID)
```
Comment (WillieCubed) [inline scraper/scraper.py:133]:
```suggestion
        downloadURL = f'https://coursebook.utdallas.edu/reportmonkey/cb11-export/{downloadID}/{downloadID}/json'
```

File: scraper/scraper.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+            filename=self.loggingFile, 
+            filemode='w', 
+            level=logging.INFO, 
+        )
+
+        # document object instantiation
+        self.logger.info('Started coursebook scraper')
+
+
+    def getSearchSpecifications(self) -> dict:
```
Comment (WillieCubed) [inline scraper/scraper.py:38]: I'll note this now: stick to PEP 8 (the Python style guide) and use that convention for all of the variable/function/class names: https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles

File: scraper/scraper.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+            'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
+            'Sec-Gpc': '1',
+            'Sec-Fetch-Site': 'same-origin',
+            'Sec-Fetch-Mode': 'cors',
+            'Sec-Fetch_dest': 'empty',
+            'Referer': 'https://coursebook.utdallas.edu/',
+        }
+
+        # send CourseBook search query
+        # self.logger.info('Sending CourseBook query: {0}'.format(query))
```
Comment (WillieCubed) [inline scraper/scraper.py:99]: Don't forget to remove this if it isn't necessary for functionality.

File: scraper/scraper.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+if __name__ == '__main__':
+    # import environment variables
+    dotenv.load_dotenv()
+
+    # instantiate scraper with ptgsessid
+    scraper = CoursebookScraper(os.environ['PTGSESSID'])
+
+    # load default config
+    with open('configs/{}'.format(os.environ['CONFIG_FILE']), 'r') as file:
```
Comment (WillieCubed) [inline scraper/scraper.py:204]:
```suggestion
    config_file_name = os.environ['CONFIG_FILE']
    with open(f'configs/{config_file_name}', 'r') as file:
```

File: scraper/scraper.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+        scraper.configuration = json.load(file)
+
+    # download data from coursebook
+    result = scraper.downloadFromConfig()
+
+    # write results to data.json
+    with open('data/{}.json'.format(int(time.time())),'w') as file:
```
Comment (WillieCubed) [inline scraper/scraper.py:211]:
```suggestion
    with open(f'data/{int(time.time()}.json', 'w') as file:
```

File: scraper/scraper.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+            html = BeautifulSoup(response.content, 'html.parser')
+        except:
+            self.logger.error('Failed to receive response from query: {0}'.format(query))
+            return {}
+
+        # parse response for items returned
+        try:
+            numItems = html.find_all('b')[0].__repr__().split('(')[1].split(')')[0].split(' ')[0] # this should probably be cleaned up
+            if numItems == 'no':
+                self.logger.info('0 results obtained from query: {0}'.format(query))
```
Comment (WillieCubed) [inline scraper/scraper.py:115]:
```suggestion
                self.logger.info(f'0 results obtained from query: {query}')
```

File: scraper/scraper.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+
+        # return obtained specifications
+        return specifications
+
+
+    def downloadCoursebookData(self, queryList: list) -> dict:
+        # format query list as a proper query string
+        query = 'action=search'
+        for specifier in queryList:
+            query += '&s[]={0}'.format(specifier)
```
Comment (WillieCubed) [inline scraper/scraper.py:81]:
```suggestion
            query += f'&s[]={specifier}'
```
Sidenote for future reference: I'm almost certain there's a library that does something like this, but this is fine.

File: scraper/scraper.py [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+        # self.logger.info('Sending CourseBook query: {0}'.format(query))
+        try:
+            response = requests.post(
+                'https://coursebook.utdallas.edu/clips/clip-cb11-hat.zog',
+                headers = coursebookHeader,
+                data = query,
+            )
+            html = BeautifulSoup(response.content, 'html.parser')
+        except:
+            self.logger.error('Failed to receive response from query: {0}'.format(query))
```
Comment (WillieCubed) [inline scraper/scraper.py:108]:
```suggestion
            self.logger.error(f'Failed to receive response from query: {query}')
```

File: scraper/scraper.py [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+
+    def downloadCoursebookData(self, queryList: list) -> dict:
+        # format query list as a proper query string
+        query = 'action=search'
+        for specifier in queryList:
+            query += '&s[]={0}'.format(specifier)
+
+        # specify CourseBook request header
+        coursebookHeader = {
+            'Cookie': 'PTGSESSID={0}'.format(self.ptgsessid),
```
Comment (WillieCubed) [inline scraper/scraper.py:85]:
```suggestion
            'Cookie': f'PTGSESSID={self.ptgsessid}',
```

File: scraper/scraper.py [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+            return {}
+
+        # parse response for items returned
+        try:
+            numItems = html.find_all('b')[0].__repr__().split('(')[1].split(')')[0].split(' ')[0] # this should probably be cleaned up
+            if numItems == 'no':
+                self.logger.info('0 results obtained from query: {0}'.format(query))
+                return {}
+        except:
+            self.logger.error('Failed to parse number of items returned from query: {0}'.format(query))
```
Comment (WillieCubed) [inline scraper/scraper.py:118]:
```suggestion
            self.logger.error(f'Failed to parse number of items returned from query: {query}')
```

File: scraper/scraper.py [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+
+        # parse response for download link
+        try:
+            downloadLink = html.find_all('a')[0]['href'].replace('\\','').replace('"','')
+            if downloadLink[0:14] == '/reportmonkey/':
+                downloadID =  downloadLink.split('/')[-1]
+            else:
+                raise Exception()
+        except:
+            self.logger.error('Failed to parse download link to {0} item(s) from query: {1}'.format(numItems, query))
```
Comment (WillieCubed) [inline scraper/scraper.py:129]:
```suggestion
            self.logger.error(f'Failed to parse download link to {numItems} item(s) from query: {query}')
```

File: scraper/scraper.py [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+            'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.81 Safari/537.36',
+        }
+
+        # send coursemonkey download request for json
+        # self.logger.info('Sending ReportMonkey query: {0}'.format(query))
+        try:
+            download = requests.get(downloadURL, headers=reportmonkeyHeader)
+            rawJSON = download.text
+        except:
+            self.logger.error('Failed to download JSON response for query: {0}'.format(query))
```
Comment (WillieCubed) [inline scraper/scraper.py:158]:
```suggestion
            self.logger.error(f'Failed to download JSON response for query: {query}')
```

File: scraper/scraper.py [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+            return {}
+
+        # convert rawJSON to JSON object
+        try:
+            queryAsJSON = json.loads(rawJSON)
+        except:
+            self.logger.error('Failed to convert JSON to dict for query: {0}'.format(query))
```
Comment (WillieCubed) [inline scraper/scraper.py:165]:
```suggestion
            self.logger.error(f'Failed to convert JSON to dict for query: {query}')
```

**Review by WillieCubed — CHANGES_REQUESTED** [review]
Some relatively minor changes to make, but nice job on the script.

**Comment by CharlieMahana** [comment]
I fixed all the issues you mentioned and removed the hardcoded data file in dataToFirestore.py so that the file can be passed into the program through command line arguments

**Review by WillieCubed — CHANGES_REQUESTED** [review]
Ah, there was a comment that was missed [here](https://github.com/UTDNebula/nebula-api/pull/22/commits/9526714b629d9ccbaa2a7a04f26c1dcae99ac451#r746878065). Additionally, add `.env` to the `.gitignore`.

**Comment by CharlieMahana** [comment]
I removed the commented code block, finished f-string conversions, and added /scraper/.env to .gitignore. My only concern is that firebase credentials could accidentally be committed to the repository as the credentials file gets a new name every time we generate one. How do we want to solve this problem?

**Comment by WillieCubed** [comment]
In the project set-up docs, just inform people to rename the file so it has the correct hardcoded name a that is how it is intended to work.

**Comment by CharlieMahana** [comment]
@WillieCubed my most recent commit should mean everything has been taken care of and we're ready to merge. Let me know if there is anything that I missed.

**Review by WillieCubed — APPROVED** [review]
LGTM!

**Pull request merged**